internal/queue/consumer.go

```diff
@@ -0,0 +1,286 @@
+	job := &database.Job{
+		ID:          strconv.Itoa(j.ID),
+		Platform:    platform,
+		Labels:      j.Labels,
+		CreatedAt:   c.parseToTime(j.CreatedAt),
+		StartedAt:   nil,
+		CompletedAt: nil,
+		Raw:         c.extractRaw(msg),
+	}
+	if err := c.db.AddJob(ctx, job); err != nil {
+		if errors.Is(err, database.ErrExist) {
+			c.logger.Info("job already exists in database", "job_id", j.ID)
+			msg.Nack(false, false) // don't requeue
+			return nil
+		}
+		msg.Nack(false, true) // requeue
+		return fmt.Errorf("failed to add job to database: %w", err)
+	}
+	msg.Ack(false)
+	return nil
+}
+
+// updateJobStartedInDB updates the job's started_at timestamp in the database.
+func (c *AmqpConsumer) updateJobStartedInDB(ctx context.Context, j WorkflowJob, msg amqp.Delivery) error {
+	if c.parseToTime(j.StartedAt).IsZero() {
+		msg.Nack(false, false) // don't requeue
+		return errors.New("started_at missing in job_started event")
+	}
+	if err := c.db.UpdateJobStarted(ctx, platform, strconv.Itoa(j.ID), c.parseToTime(j.StartedAt), c.extractRaw(msg)); err != nil {
+		if errors.Is(err, database.ErrNotExist) {
+			c.logger.Warn("job not found in database on start", "job_id", j.ID)
+			msg.Nack(false, false) // don't requeue
+			return nil
+		}
+		msg.Nack(false, true) // requeue
+		return fmt.Errorf("failed to update job started in database: %w", err)
+	}
+	msg.Ack(false)
+	return nil
+}
+
+// updateJobCompletedInDB updates the job's completed_at timestamp in the database.
+func (c *AmqpConsumer) updateJobCompletedInDB(ctx context.Context, j WorkflowJob, msg amqp.Delivery) error {
+	if c.parseToTime(j.CompletedAt).IsZero() {
+		msg.Nack(false, false) // don't requeue
+		return errors.New("completed_at missing in job_completed event")
+	}
+	if err := c.db.UpdateJobCompleted(ctx, platform, strconv.Itoa(j.ID), c.parseToTime(j.CompletedAt), c.extractRaw(msg)); err != nil {
+		if errors.Is(err, database.ErrNotExist) {
+			c.logger.Warn("job not found in database on completion", "job_id", j.ID)
+			msg.Nack(false, false) // don't requeue
+			return nil
+		}
+		msg.Nack(false, true) // requeue
+		return fmt.Errorf("failed to update job completed in database: %w", err)
+	}
+	msg.Ack(false)
+	return nil
+}
+
+// extractRaw extracts the raw payload from the AMQP message.
+func (c *AmqpConsumer) extractRaw(msg amqp.Delivery) map[string]any {
+	var payload map[string]any
+	if err := json.Unmarshal(msg.Body, &payload); err != nil {
+		c.logger.Warn("failed to extract raw payload", "error", err)
+		return nil
+	}
+	return payload
+}
+
+// consumeMsgs starts consuming messages from the AMQP queue.
+func (c *AmqpConsumer) consumeMsgs() (<-chan amqp.Delivery, error) {
+	msgs, err := c.amqpChannel.Consume(
+		c.queueName, // queue
+		"",          // consumer tag, empty string means a unique random tag will be generated
+		false,       // whether rabbitmq auto-acknowledges messages
+		false,       // whether only this consumer can access the queue
```

Comment:
This will have consequences if/when we'll scale up the planner. This flag will allow only 1 unit to consume the events when set to trie, while if set to false, the events will be distributed.
If we don't plan to scale up, I'd rather put this flag to true to avoid potential race conditions
Otherwise, we need to pay attention to race conditions

Reply:
I suggest implementing true for now (don't allow scale-up usecase), and we discuss in design meeting for our long term approach.